src/jpeg/encoder.rs

```diff
@@ -771,82 +788,84 @@ fn encode_coefficient(coefficient: i32) -> (u8, u16) {
     (num_bits, val)
 }
 
-fn rgb_to_ycbcr(r: u8, g: u8, b: u8) -> (u8, u8, u8) {
-    let r = f32::from(r);
-    let g = f32::from(g);
-    let b = f32::from(b);
+#[inline]
+fn rgb_to_ycbcr<P: Pixel>(pixel: P) -> (u8, u8, u8) {
+    use num_traits::{cast::ToPrimitive, bounds::Bounded};
+    let [r, g, b] = pixel.to_rgb().0;
+    let max: f32 = P::Subpixel::max_value().to_f32().unwrap();
+    let r: f32 = r.to_f32().unwrap();
+    let g: f32 = g.to_f32().unwrap();
+    let b: f32 = b.to_f32().unwrap();
 
-    let y = 0.299f32 * r + 0.587f32 * g + 0.114f32 * b;
-    let cb = -0.1687f32 * r - 0.3313f32 * g + 0.5f32 * b + 128f32;
-    let cr = 0.5f32 * r - 0.4187f32 * g - 0.0813f32 * b + 128f32;
+    let y = 65.481 / max * r + 128.553 / max * g + 24.933 / max * b;
+    let cb = -37.797 / max * r - 74.203 / max * g + 112.0 / max * b + 128.;
+    let cr = 112. / max * r - 93.786 / max * g - 18.214 / max * b + 128.;
 
     (y as u8, cb as u8, cr as u8)
 }
 
-fn value_at(s: &[u8], index: usize) -> u8 {
-    if index < s.len() {
-        s[index]
+
+/// Returns the pixel at (x,y) if (x,y) is in the image,
+/// otherwise the closest pixel in the image
+#[inline]
+fn pixel_at_or_near<I: GenericImageView>(source: &I, x: u32, y: u32) -> I::Pixel {
+    if source.in_bounds(x, y) {
+        unsafe { source.unsafe_get_pixel(x, y) }
```

Comment:
Did you try it without resorting to `unsafe` here? If so, how bad was the performance regression? If not, why not and can you try it out?

Reply:
Yes I removed the redundant bound checks because there was a ~8% performance difference. This function is called for every pixel in the image.

Reply:
`GenericImageView::get_pixel` should probably return an option by default, but this is another discussion. The way it works right now, it forces duplicated bound checks.

Reply:
I'm slightly uneasy about this. Maybe `ImageBuffer` could specialize the `in_bounds` method to help the optimizer? In any case I am reminded of similar bounds checking problems in `imageproc` and a general solution would be nice. An alternative would be to have different implementations for the general case but I'm not sure its preferrable.

@fintelia I would be interested in your input here. You are always more cautious about `unsafe` in the first place.

Reply:
I can move back to the double bound checks and remove the unsafe if it can help to move more quickly. We can always change that later.

Reply:
`GenericImageView::get_pixel` returning an Option does sound like the cleanest solution here. It would indeed avoid the duplicate bounds checks, and I think it would not require any `unsafe` either because can delegate to the pre-existing methods on slices etc.

Reply:
Do you want me to make the change as part of this pull request ?

Reply:
A separate PR would be better, it could be completed in parallel if you want to.

Reply:
OK. I removed the unsafe_get_pixel from this PR. Let me know if there is anything more to do before it can be merged.